3단계 - 로또(자동)

daehyeon0 opened the pull request:
늦은 밤에 PR 드리게 되어 죄송스럽습니다.

2단계 때 생각보다 잘된다고 했었는데.. 그냥 요구사항이 쉬워서 그랬던 것 같습니다..
3단계 리뷰 잘부탁드립니다!

daehyeon0 committed:
docs: 요구사항 추가

daehyeon0 committed:
docs: 기능 요구사항 변경

- 번호 생성의 책임을 LottoGame으로 이동
- LottoNumbers는 고정된 길이의 LottoNumber 주입 생성자를 갖도록 수정

daehyeon0 committed:
feat: LottoNumber 구현

daehyeon0 committed:
feat: LottoNumbers 구현

daehyeon0 committed:
docs: LottoNumbers 기능 요구사항 추가 및 수정

daehyeon0 committed:
feat: LottoNumbers 생성자 추가

daehyeon0 committed:
feat: Lottonumbers 숫자가 6개가 아닌 경우 예외 발생 추가

daehyeon0 committed:
feat: Lottonumbers 숫자에 겹치는 숫자가 있을 경우 예외

daehyeon0 committed:
docs: 오타 제거

daehyeon0 committed:
docs: 기능 요구사항 추가

- InputView 클래스 기능 추가

daehyeon0 committed:
feat: InputView 클래스 구현

daehyeon0 committed:
chore: package 이동

daehyeon0 committed:
test: package 이동에 따라 패키지 import

daehyeon0 committed:
feat: LottoPurchaseCount 클래스 구현

daehyeon0 committed:
chore: package 이동

- calculator package 추가 및 기존 클래스 이동

daehyeon0 committed:
refactor: CalculatorInputStringParser 기능 일부 변경

- split한 요소들의 공백을 제거하고 Integer로 parse

daehyeon0 committed:
feat: LottoNumbers 생성자 추가

- delimiter로 구분되는 문자열을 이용한 생성 가능

daehyeon0 committed:
feat: LottoNumberChoiceRandom 클래스 구현

daehyeon0 committed:
docs: 기능 요구사항 추가

- LottoNumberChoiceStrategy 관련한 기능 추가

daehyeon0 committed:
docs: 기능 구현목록 수정

daehyeon0 committed:
feat: Rank 추가

daehyeon0 committed:
feat: RewardCalculator 추가

daehyeon0 committed:
feat: LottoNumber에 getNumber 메서드 추가

daehyeon0 committed:
feat: LottoNumberChoiceStrategy 관련 기능 추가

- 구체 클래스 LottoNumberChoiceRandom 추가

daehyeon0 committed:
feat: LottoResult 추가

daehyeon0 committed:
feat: 다른 Lotto와 비교하는 기능 추가

- LottoNumbers -> Lotto로 클래스 변경
- Strategy를 이용하여 클래스를 생성하는 생성자 추가

daehyeon0 committed:
fix: LottoPurchaseCount 버그 수정

- 생성자에서 입력받는 값을 횟수 -> 가격으로 수정

daehyeon0 committed:
feat: Lottos 추가

daehyeon0 committed:
feat: inputView 기능 추가

- 구매한 로또들을 보여주는 기능 추가

daehyeon0 committed:
feat: ResultView 추가

- 로또 게임 통계에 관련된 View 클래스 추가

daehyeon0 committed:
feat: LottoGame 추가

daehyeon0 committed:
build: 프로그램 엔트리 추가

daehyeon0 committed:
chore: CalculatorInputStringParser의 클래스 이름 변경

daehyeon0 committed:
feat: inputView 기능 추가

- 사용자의 입력 예외상황을 가정한 메세지 추가

daehyeon0 committed:
feat: 잘못된 입력을 받을 경우 다시 입력받도록 수정

daehyeon0 committed:
feat: LottoNumber 생성 시 숫자가 로또 범위내에 있는지 체크

daehyeon0 committed:
style: 빈 행 삭제

jaeyeonling commented on src/main/java/controller/LottoGame.java:
```diff
@@ -0,0 +1,67 @@
+
+public class LottoGame {
+	InputView inputView = new InputView();
+	ResultView resultView = new ResultView();
```
```suggestion
	private final InputView inputView = new InputView();
	private final ResultView resultView = new ResultView();
```

접근제어자를 붙여주는 습관을 들여보아요!

jaeyeonling commented on src/main/java/exception/OutOfRange.java:
```diff
@@ -0,0 +1,7 @@
+package exception;
+
+public class OutOfRange extends RuntimeException {
```
```suggestion
public class OutOfRangeException extends RuntimeException {
```

Exception이라는 suffix를 붙여줄 필요는 없을까요?

daehyeon0 replied:
앗.. 좀 더 명확히 알 수 있도록 추가하겠습니다!

jaeyeonling commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+		}
+		validateDuplication(lottoNumbers);
+
+		this.lottoNumbers = new ArrayList<>();
+		sortLottoNumbers(lottoNumbers);
+		this.lottoNumbers.addAll(lottoNumbers);
```
```suggestion
		this.lottoNumbers = new ArrayList<>(lottoNumbers);
		sortLottoNumbers(this.lottoNumbers);
```

소소하지만 생성자를 통해 초기화해볼 수 있을 것 같네요~

jaeyeonling commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+		lottoNumbers.sort(LottoNumber::getComparatorOther);
+	}
+
+	private void validateDuplication(List<LottoNumber> lottoNumbers) {
+		for (int i = 0; i < lottoNumbers.size(); ++i) {
+			validateDuplication(lottoNumbers, i);
+		}
+	}
+
+	private void validateDuplication(List<LottoNumber> lottoNumbers, int i) {
+		for (int j = i + 1; j < lottoNumbers.size(); ++j) {
+			validateDuplication(lottoNumbers, i, j);
+		}
+	}
+
+	private void validateDuplication(List<LottoNumber> lottoNumbers, int i, int j) {
+		if (lottoNumbers.get(i).equals(lottoNumbers.get(j))) {
+			throw new IllegalArgumentException(MESSAGE_NOT_ALLOW_DUPLICATION);
+		}
+	}
```
중복을 허용하지 않는 Set 자료구조를 통해 조금 더 간결하게 만들어볼 수 있을 것 같네요~

daehyeon0 replied:
중복 허용 여부를 검사하기위해 Set을 사용해보겠습니다!

jaeyeonling commented on src/main/java/model/Lotto.java:
```diff
@@ -0,0 +1,104 @@
+	}
+
+	@Override
+	public String toString() {
```
LottoNumber.toString을 구현한다면 해당 toString도 간결하게 작성할 수 있을 것 같네요!

daehyeon0 replied:
LottoNumber.toString을 구현하는 이유가 getNumber보다는 내부 구현을 덜 노출하기 때문이라고 생각하는데요,
LottoNumber.toString을 구현함으로써 해당 코드를 간결하게 작성할 수도 있을까요?

jaeyeonling replied:
LottoNumber.toString를 구현한다면 lottoNumbers.toString() 를 통해 동일한 결과를 볼 수 있을 것 같아요 😄

jaeyeonling commented on src/main/java/model/Rank.java:
```diff
@@ -0,0 +1,21 @@
+
+public enum Rank {
+	FIRST(2000000000),
+	SECOND(0),
```
SECOND 가 0인 이유가 있을까요?

daehyeon0 replied:
이번 Step에서는 정의될 부분이 아닌 것 같아서 0으로 지정 해 뒀는데..
버그로 SECOND가 추가되어도 합산 통계에는 영향을 끼치지 않도록요.
생각해보니 그럴거면 아예 enum이 정의되지 않는 편이 나을까요?

jaeyeonling commented on src/main/java/model/RewardCalculator.java:
```diff
@@ -0,0 +1,47 @@
+public class RewardCalculator {
+	public static final int DEFAULT_COUNT = 0;
+
+	private final Map<Rank, Integer> rankMap;
```
Map과 같은 자료형을 이름으로 사용하면 어떠한 문제가 있을지 고민해보아요

daehyeon0 replied:
자료형에 종속적이게 되어서 자료형을 변경할 때 같이 변경해야하는 점일까요?

jaeyeonling commented on src/test/java/LottoTest.java:
```diff
@@ -0,0 +1,102 @@
+	@Test
+	@DisplayName("로또 숫자가 6개가 포함되면서 겹치는 값이 없으면 성공")
+	void test_constructor3() {
+		assertThatNoException().isThrownBy(() -> new Lotto(
```
해당 assert는 아래 assert로 같이 검증이 가능할 것 같네요~
또한 매번 `new LottoNumber`를 반복하는 것은 매우 힘든 일이기 때문에 `new Lotto(Arrays.asList(1, 2, 3, 4, 5, 6))` 혹은 `new Lotto(1, 2, 3, 4, 5, 6)`와 같은 생성자를 제공해주면 어떨까요?

daehyeon0 replied:
넵!

jaeyeonling commented on src/test/java/LottosTest.java:
```diff
@@ -0,0 +1,52 @@
+import model.Rank;
+import model.RewardCalculator;
+
+public class LottosTest {
```
해당 도메인이 가장 핵심이 되는 도메인으로 보여지는데, 생각보다 테스트가 부족하네요!
여러 케이스에 대한 당첨 결과를 검증해보면 좋을 것 같아요 :)

daehyeon0 replied:
음... 일단 Lottos 생성자 내부의 Lotto와 LottoPurchaseCount의 예외를 가져와서 테스트 케이스에 추가하긴 했는데요..
Lotto와 LottoPurchaseCount에서 해당 테스트를 실행하고 있는데 LottosTest에서도 테스트 할 필요가 있는지에 대해 궁금합니다!
작성하면서 만약 여러 모듈에 의존하는 모듈(현재의 LottosTest)이라면 여러 모듈을 엮어서 또 다른 분기가 나올 수 있겠다는 생각은 드는데.. 그럼 추가된 분기에 대해 테스트하면 되는 것 아닌가요?

jaeyeonling replied:
작성하신 테스트 이외에 등수에 따른 테스트를 조금 더 탄탄히 작성해주시면 좋을 것 같아요 ㅎㅎ 
말씀해주신 부분도 굉장히 고민이 되는 부분 중 하나인데요
테스트 코드를 작성하다보면 중복되는 테스트가 굉장히 많이 나오게 되는데, 저는 개인적으로 복붙을 해서 중복되는 테스트를 작성합니다
이유는 해당 클래스의 설명서 역할을 할 수 있고, 해당 클래스에 테스트가 없는 상태에서 리팩토링될 경우 예상치 못하는 문제를 잡기 위함입니다!

jaeyeonling reviewed (changes requested):
로또 자동을 잘 구현해주셨습니다 👍 
새벽에도 편하게 PR 주셔도 됩니다 😄 
코멘트 남겨두었으니 확인 부탁드릴게요~

daehyeon0 commented:
정말 많이 노력해야겠다는 생각이 드네요.
남은 기간 동안 꾸준히 시도해보겠습니다.
세심한 리뷰 감사합니다.

daehyeon0 committed:
refactor: 멤버 변수에 불변 및 접근제어자 추가

- LottoGame.inputView, LottoGame.resultView 변수에 해당 내용 적용

daehyeon0 committed:
refactor: 명확하게 Exception로 표기하기 위한 클래스 이름 변경

- OutOfRange -> OutOfRangeException

daehyeon0 committed:
refactor: 코드 가독성을 위해 더 짧은 코드로 변경

- 별도의 함수를 이용해 초기화 하던 것을 생성자를 통한 초기화로 변경

daehyeon0 committed:
refactor(Lotto): 중복 검사 코드 단축

- for문을 돌며 contain을 사용하던 것을 Set을 이용하도록 수정

daehyeon0 committed:
refactor: LottoNumber의 toString 구현 및 Lotto에서 사용

daehyeon0 committed:
refactor(Rank): 현재 필요하지 않은 항목 제거

daehyeon0 committed:
refactor(RewardCalculator): 멤버변수명 변경

- rankMap -> rankMatchingCounts로 변경

daehyeon0 committed:
test(LottoTest): 필요없는 테스트 삭제

daehyeon0 committed:
refactor: 클래스 생성자 변경

- Lotto 클래스의 생성자로 List<LottoNmber>을 제공해야 하던 것을
  List<Integer>로 전달하고 내부에서 LottoNumber로 생성하는 것으로 변경

daehyeon0 committed:
test: 잘못 작성된 test 변경

- Rank.SECOND를 제거했음에도 불구하고 테스트에 남겨둔 것 삭제

daehyeon0 committed:
test: 필요없는 테스트 제거

- LottoTest의 필요없는 테스트를 제거

daehyeon0 committed:
docs: 기능 요구사항 추가

- LottoPurchaseCount 생성자 파라미터로 0원을 제공할 경우 예외 추가

daehyeon0 committed:
feat: LottoPurchaseCount 생성자 예외 케이스 추가

- 0원을 제공하면 예외

daehyeon0 committed:
test: LottosTest 추가

jaeyeonling commented on src/main/java/model/Rank.java:
```diff
@@ -0,0 +1,20 @@
+package model;
+
+public enum Rank {
```
Q. 이번 Step에서는 정의될 부분이 아닌 것 같아서 0으로 지정 해 뒀는데..
버그로 SECOND가 추가되어도 합산 통계에는 영향을 끼치지 않도록요.
생각해보니 그럴거면 아예 enum이 정의되지 않는 편이 나을까요?

A. 맞습니다~ 
원래는 단계별로 진행되기 때문에 뒷 요구사항은 모르는 상태로 진행된다고 생각해주시면 될 것 같아요 😄

jaeyeonling commented on src/main/java/model/RewardCalculator.java:
```diff
@@ -0,0 +1,47 @@
+public class RewardCalculator {
+	public static final int DEFAULT_COUNT = 0;
+
+	private final Map<Rank, Integer> rankMatchingCounts;
```
Q. 자료형에 종속적이게 되어서 자료형을 변경할 때 같이 변경해야하는 점일까요?
A. 맞습니다 😄

jaeyeonling commented on src/main/java/model/LottoResult.java:
```diff
@@ -0,0 +1,47 @@
+		matchingCount++;
+	}
+
+	public Rank getRank() {
```
Rank에서 matcingCount를 가지면 조금 더 간단하게 처리 할 수 있겠네요 ㅎㅎ

jaeyeonling reviewed (approved):
코멘트를 잘 구현해주셨습니다 👍 
질문의 답변과 추가적인 코멘트 남겨두었으니 확인 부탁드릴게요~ 
확인 후 다음 단계 반영해주세요!!
